### DISCUSSION TITLE
How do I access Pulsar devservices in image based integration tests?

### DISCUSSION THREAD

**Discussion by jeloba** [quarkusio/quarkus#45177] [Q&A]
When starting an integration test for testing an image artifact, both the app and the devservice container will be put inside a docker network. While the app exposes it's ports on the host, the devservice (Pulsar in my case) is not accessible.

For my tests, I need to access the devservice, both to set up parts of the test environment, and to send and receive test messages.

How do I access a devservice running inside a docker network form the integration test?

For context, I'm running the app with

```./mvnw clean verify -DskipITs=false -Dquarkus.test.integration-test-profile=test-with-native-agent```

[a comment by an automated account is not shown]

**Reply by geoand** [reply]
cc @ozangunalp

**Reply by jeloba** [reply]
@geoand @ozangunalp any advice on how to proceed?

**Comment by jeloba** [comment]
Just gave it a try with running on the host network (`quarkus.test.container.network=host`) but this time the application failed to start. Sees like the Pulsar devservice ran in host mode, while the app container was started on a different network.

**Reply by jeloba** [reply]
Looking through the code determining the network used for spawned containers, I came across this block [here](https://github.com/quarkusio/quarkus/blob/9943ebd10a289dd878b00441c88be91ab62afb92/test-framework/junit5/src/main/java/io/quarkus/test/junit/IntegrationTestUtil.java#L267C1-L286C14)

```java
        if (isDockerAppLaunch) {
            // obtain the ID of the shared network - this needs to be done after the augmentation has been run
            // or else we run into various ClassLoader problems
            try {
                Class<?> networkClass = curatedApplication.getOrCreateAugmentClassLoader()
                        .loadClass("org.testcontainers.containers.Network");
                Object sharedNetwork = networkClass.getField("SHARED").get(null);
                networkId = (String) networkClass.getMethod("getId").invoke(sharedNetwork);
            } catch (Exception e) {
                // use the network the use has specified or else just generate one if none is configured


                Optional<String> networkIdOpt = ConfigProvider.getConfig().getOptionalValue("quarkus.test.container.network",
                        String.class);
                if (networkIdOpt.isPresent()) {
                    networkId = networkIdOpt.get();
                } else {
                    networkId = "quarkus-integration-test-" + RandomStringUtils.insecure().next(5, true, false);
                    manageNetwork = true;
                }
            }
        }
```

Shouldn't the config `quarkus.test.container.network` override the `SHARED` test-containers network, instead of the other way round?

At least to me it looks like the Pulsar devservice is respecting the setting, while the app container is not.
